domain/application/service/resource.go

```diff
@@ -0,0 +1,226 @@
+type ResourceState interface {
+	// GetApplicationResourceID returns the ID of the application resource
+	// specified by natural key of application and resource name.
+	GetApplicationResourceID(ctx context.Context, args resource.GetApplicationResourceIDArgs) (resources.ID, error)
+
+	// ListResources returns the list of resources for the given application.
+	ListResources(ctx context.Context, applicationID application.ID) (resource.ApplicationResources, error)
+
+	// GetResource returns the identified resource.
+	GetResource(ctx context.Context, resourceID resources.ID) (resource.Resource, error)
+
+	// SetResource adds the resource to blob storage and updates the metadata.
+	SetResource(ctx context.Context, config resource.SetResourceArgs) (resource.Resource, error)
+
+	// SetUnitResource sets the resource metadata for a specific unit.
+	SetUnitResource(ctx context.Context, config resource.SetUnitResourceArgs) (resource.SetUnitResourceResult, error)
+
+	// OpenResource returns the metadata for a resource and a reader for
+	// the resource.
+	OpenResource(ctx context.Context, resourceID resources.ID) (resource.Resource, io.ReadCloser, error)
+	// OpenApplicationResource?
+
+	// OpenResourceForUniter returns the metadata for a resource and a reader
+	// for the resource. A unit resource is created to track the given unit and
+	// which resource its using.
+	OpenResourceForUniter(ctx context.Context, resourceID resources.ID, unitID coreunit.ID) (resource.Resource, io.ReadCloser, error)
+	// OpenUnitResource?
+
+	// SetRepositoryResources sets the "polled" resources for the
+	// application to the provided values. The current data for this
+	// application/resource combination will be overwritten.
+	SetRepositoryResources(ctx context.Context, config resource.SetRepositoryResourcesArgs) error
+}
+
+// ResourceService provides the API for working with resources.
+type ResourceService struct {
+	st     ResourceState
+	logger logger.Logger
+}
+
+// NewResourceService returns a new service reference wrapping the input state.
+func NewResourceService(st ResourceState, logger logger.Logger) *ResourceService {
+	return &ResourceService{
+		st:     st,
+		logger: logger,
+	}
+}
+
+// GetApplicationResourceID returns the ID of the application resource specified by
+// natural key of application and resource name.
+//
+// The following error types can be expected to be returned:
+//   - application.ResourceNotValid if no resource name is provided in
+//     the args.
+//   - errors.NotValid is returned if the application ID is not valid.
+//   - application.ResourceNotFound if no resource with name exists for
+//     given application.
+func (s *ResourceService) GetApplicationResourceID(ctx context.Context, args resource.GetApplicationResourceIDArgs) (resources.ID, error) {
+	if err := args.ApplicationID.Validate(); err != nil {
+		return "", fmt.Errorf("application id: %w", err)
+	}
+	if args.Name == "" {
+		return "", applicationerrors.ResourceNotValid
```

Comment:
Either we change the error to be `ResourceNotValid` or add context about why the name isn't valid.